web/modules/sheaf/src/components/MorphingWikitext.svelte

```diff
@@ -0,0 +1,105 @@
+<!--
+  @component Generic wikitext container.
+-->
+<script lang="ts">
+  import FTML, { type RenderedHTML } from "@wikijump/ftml-wasm-worker"
+  import {
+    animationFrame,
+    createLock,
+    createMutatingLock,
+    idleCallback,
+    perfy,
+    toFragment
+  } from "@wikijump/util"
+  import * as micromorph from "micromorph"
+  import type { SheafState } from "../state"
+
+  type Wikitext = SheafState | string
+
+  export let wikitext: Wikitext = ""
+
+  export let rendering = false
+
+  export let timeTotal = 0
+  export let timeCompile = 0
+  export let timePatch = 0
+
+  let element: HTMLElement
+  let stylesheets: string[] = []
+  let measureTotal: (() => number) | null = null
+
+  function stylesheet(node: HTMLStyleElement, style: string) {
+    node.innerHTML = style
+    return {
+      update(style: string) {
+        node.innerHTML = style
+      },
+      destroy() {
+        // not sure why I have to destroy the node, but I do apparently
+        node.parentElement?.removeChild(node)
+      }
+    }
+  }
+
+  const compile = createMutatingLock(async (wikitext: Wikitext) => {
+    rendering = true
+    const measureCompile = perfy()
+    if (!measureTotal) measureTotal = perfy()
+
+    const result =
+      typeof wikitext === "string"
+        ? await FTML.renderHTML(wikitext, undefined, "draft")
+        : await wikitext.render()
+
+    timeCompile = measureCompile()
+    return result
+  })
+
+  const patch = createLock(async ({ html, styles }: RenderedHTML) => {
```

Comment:
For my information, what does the `createLock` utility do?

Reply:
Spits out a function that when called multiple times, will wait for each previous call to finish. Basically, only one instance of the function can be active at a time.